core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,151 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static java.util.Objects.requireNonNull;
+
+public class ArrayAggregationAfterDistinct
+        implements Rule<ProjectNode>
+{
+    private static final Capture<AggregationNode> CHILD = newCapture();
+    private final Metadata metadata;
+
+    public ArrayAggregationAfterDistinct(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return project()
+                .matching(entry -> hasArrayDistinct(metadata, entry))
+                .with(source().matching(aggregation()
+                        .matching(ArrayAggregationAfterDistinct::hasArrayAggregation)
+                        .capturedAs(CHILD)));
+    }
+
+    private static boolean hasArrayAggregation(AggregationNode aggregationNode)
+    {
+        return aggregationNode.getAggregations().values().stream()
+                .anyMatch(aggregation -> aggregation.getResolvedFunction().getSignature().getName().equals(ArrayAggregationFunction.NAME) && aggregation.getArguments().size() == 1);
+    }
+
+    private static boolean hasArrayDistinct(Metadata metadata, ProjectNode projectNode)
+    {
+        return projectNode.getAssignments().getExpressions().stream()
+                .anyMatch(assignment -> assignment instanceof FunctionCall &&
+                        metadata.decodeFunction(((FunctionCall) assignment).getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME));
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        AggregationNode aggregationNode = captures.get(CHILD);
+
+        Assignments.Builder assignmentsBuilder = Assignments.builder();
+        ImmutableMap.Builder<Symbol, Aggregation> aggregationsBuilder = ImmutableMap.builder();
+
+        List<Symbol> matchedAggregations = Lists.newArrayList();
+        List<Symbol> matchedAssignments = Lists.newArrayList();
+
+        for (Map.Entry<Symbol, Expression> entry : projectNode.getAssignments().entrySet()) {
+            if (entry.getValue() instanceof FunctionCall) {
+                Symbol symbol = entry.getKey();
+                FunctionCall functionCall = (FunctionCall) entry.getValue();
+                ResolvedFunction resolvedFunction = metadata.decodeFunction(functionCall.getName());
+                if (resolvedFunction.getSignature().getName().equals(ArrayDistinctFunction.NAME)) {
+                    SymbolReference argument = (SymbolReference) (getOnlyElement(functionCall.getArguments()));
```

Comment:
Do not assume that this is a `SymbolReference`. This could be arbitrary Expression.